domain/domain.go

```diff
@@ -1224,6 +1252,180 @@ func (do *Domain) NotifyUpdatePrivilege(ctx sessionctx.Context) {
+	return atomic.LoadUint64(&do.serverID)
+}
+
+// IsLostConnectionToPD indicates lost connection to PD or not.
+func (do *Domain) IsLostConnectionToPD() bool {
+	return do.isLostConnectionToPD.Get() != 0
+}
+
+const (
+	serverIDEtcdPath = "/tidb/server_id"
+	// serverIDTTL should be LONG ENOUGH to avoid barbarically killing an on-going long-run SQL.
+	serverIDTTL = 12 * time.Hour
+	// serverIDTimeToCheckPDConnection is the interval that we check connection to PD periodically.
+	serverIDTimeToCheckPDConnection = 5 * time.Minute
+	// lostConnectionToPDTimeout is the duration that when TiDB cannot connect to PD excceeds this limit,
+	//   we realize the connection to PD is lost utterly, and server ID acquired before should be released.
+	//   Must be SHORTER than `serverIDTTL`.
+	lostConnectionToPDTimeout = 6 * time.Hour
+
+	refreshServerIDRetryCnt        = 3
+	acquireServerIDRetryInterval   = 300 * time.Millisecond
+	acquireServerIDTimeout         = 10 * time.Second
+	retrieveServerIDSessionTimeout = 10 * time.Second
+)
+
+func (do *Domain) retrieveServerIDSession(ctx context.Context) (*concurrency.Session, error) {
+	if do.serverIDSession != nil {
+		return do.serverIDSession, nil
+	}
+
+	// `etcdClient.Grant` needs a shortterm timeout, to avoid blocking if connection to PD lost,
+	//   while `etcdClient.KeepAlive` should be longterm.
+	//   So we separately invoke `etcdClient.Grant` and `concurrency.NewSession` with leaseID.
+	childCtx, cancel := context.WithTimeout(ctx, retrieveServerIDSessionTimeout)
+	resp, err := do.etcdClient.Grant(childCtx, int64(serverIDTTL.Seconds()))
+	cancel()
+	if err != nil {
+		logutil.BgLogger().Error("retrieveServerIDSession.Grant fail", zap.Error(err))
+		return nil, err
+	}
+	leaseID := resp.ID
+
+	session, err := concurrency.NewSession(do.etcdClient,
+		concurrency.WithLease(leaseID), concurrency.WithContext(context.Background()))
+	if err != nil {
+		logutil.BgLogger().Error("retrieveServerIDSession.NewSession fail", zap.Error(err))
+		return nil, err
+	}
+	do.serverIDSession = session
+	return session, nil
+}
+
+func (do *Domain) acquireServerID(ctx context.Context) error {
+	atomic.StoreUint64(&do.serverID, 0)
+
+	session, err := do.retrieveServerIDSession(ctx)
+	if err != nil {
+		return err
+	}
+
+	for {
+		randServerID := rand.Int63n(int64(util.MaxServerID)) + 1 // get a random serverID: [1, MaxServerID]
+		key := fmt.Sprintf("%s/%v", serverIDEtcdPath, randServerID)
+		cmp := clientv3.Compare(clientv3.CreateRevision(key), "=", 0)
+		value := "0"
+
+		childCtx, cancel := context.WithTimeout(ctx, acquireServerIDTimeout)
+		txn := do.etcdClient.Txn(childCtx)
+		t := txn.If(cmp)
+		resp, err := t.Then(clientv3.OpPut(key, value, clientv3.WithLease(session.Lease()))).Commit()
+		cancel()
+		if err != nil {
+			return err
+		}
+		if !resp.Succeeded {
+			logutil.BgLogger().Info("random serverID exists, try again", zap.Int64("randServerID", randServerID))
+			time.Sleep(acquireServerIDRetryInterval)
```

Comment:
Is this sleep necessary? Should the server acquire a unique server ID ASAP?

Reply:
I think it is necessary, to avoid invoking too much requests to PD.

As we have 2^23 = 800w+ server ids in total, a random server id exists in PD should be quite unlikely. So I think it would be better to retry with caution.